repository/src/main/java/org/alfresco/repo/search/impl/AbstractCategoryServiceImpl.java

```diff
@@ -1,623 +1,677 @@
+    {
+        return dictionaryService.getSubAspects(ContentModel.ASPECT_CLASSIFIABLE, true);
+    }
+
+    public NodeRef createClassification(StoreRef storeRef, QName typeName, String attributeName)
+    {
+        throw new UnsupportedOperationException();
+    }
+
+    public PagingResults<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName)
+    {
+        return getRootCategories(storeRef, aspectName, pagingRequest, sortByName, null, null);
+    }
+
+    public PagingResults<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName, String filter)
+    {
+        final Collection<String> alikeNamesFilter = Optional.ofNullable(filter).map(f -> "*".concat(f).concat("*")).map(Set::of).orElse(null);
+        return getRootCategories(storeRef, aspectName, pagingRequest, sortByName, null, alikeNamesFilter);
+    }
+
+    public PagingResults<ChildAssociationRef> getRootCategories(StoreRef storeRef, QName aspectName, PagingRequest pagingRequest, boolean sortByName,
+        Collection<String> exactNamesFilter, Collection<String> alikeNamesFilter)
+    {
+        final Set<NodeRef> nodeRefs = getClassificationNodes(storeRef, aspectName);
+        final List<ChildAssociationRef> associations = new LinkedList<>();
+        final int skipCount = pagingRequest.getSkipCount();
+        final int maxItems = pagingRequest.getMaxItems();
+        final int size = (maxItems == CannedQueryPageDetails.DEFAULT_PAGE_SIZE ? CannedQueryPageDetails.DEFAULT_PAGE_SIZE : skipCount + maxItems);
+        int count = 0;
+        boolean moreItems = false;
+
+        final Function<NodeRef, Collection<ChildAssociationRef>> childNodesSupplier = (nodeRef) -> {
+            final List<ChildAssociationRef> childNodes = new ArrayList<>();
+            if (CollectionUtils.isEmpty(exactNamesFilter) && CollectionUtils.isEmpty(alikeNamesFilter))
+            {
+                // lookup in DB without filtering
+                childNodes.addAll(nodeService.getChildAssocs(nodeRef, ContentModel.ASSOC_SUBCATEGORIES, RegexQNamePattern.MATCH_ALL));
+            }
+            else
+            {
+                if (CollectionUtils.isNotEmpty(exactNamesFilter))
+                {
+                    // lookup in DB filtering by name
+                    childNodes.addAll(nodeService.getChildrenByName(nodeRef, ContentModel.ASSOC_SUBCATEGORIES, exactNamesFilter));
+                }
+                if (CollectionUtils.isNotEmpty(alikeNamesFilter))
+                {
+                    // lookup using search engin filtering by name
+                    childNodes.addAll(getChildren(nodeRef, Mode.SUB_CATEGORIES, Depth.IMMEDIATE, sortByName, alikeNamesFilter, skipCount + maxItems + 1));
+                }
+            }
+
+            Stream<ChildAssociationRef> childNodesStream = childNodes.stream().distinct();
```

Comment:
I think I mentioned this before, but if we used a Set then we wouldn't need to use `distinct` here.

Reply:
Will fix it.